### PR TITLE
Fix memory leak in UnauthenticatedSessionTable.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
UnauthenticatedSessionTable essentially assumed that non-heap pools were used and was:

1) Never releasing its entries back to the pool.
2) Assuming that the pool would fill up and then its "reuse already allocated
   entry with zero refcount" code would kick in.

Since heap pools never fill up, this meant that every single UnauthenticatedSession allocated was leaked.  And we had a helpful "release them all on destruction" to cover up the leak at shutdown and prevent leak tools from finding it.

This fix:

* Preserves existing behavior for non-heap pools.
* Switches to releasing UnauthenticatedSessions back to the pool in the heap case.

### CODE REVIEW

File: src/transport/UnauthenticatedSessionTable.h
Comment (tcarmelveilleux) [inline src/transport/UnauthenticatedSessionTable.h:137]: Tentatively wait for approval from @kpschoedel

Comment (kpschoedel) [reply src/transport/UnauthenticatedSessionTable.h:137]: Looks fine. Not thrilled by having more `#if`s, but I can't think of anything better here.

**Review by kpschoedel — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**
